framework/src/utils/LibtorchArtificialNeuralNet.C

```diff
@@ -0,0 +1,171 @@
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifdef LIBTORCH_ENABLED
+#include "LibtorchArtificialNeuralNet.h"
+#include "MooseError.h"
+
+namespace Moose
+{
+
+LibtorchArtificialNeuralNet::LibtorchArtificialNeuralNet(
+    const std::string name,
+    const unsigned int num_inputs,
+    const unsigned int num_outputs,
+    const std::vector<unsigned int> & num_neurons_per_layer,
+    const std::vector<std::string> activation_function)
+  : _name(name),
+    _num_inputs(num_inputs),
+    _num_outputs(num_outputs),
+    _num_neurons_per_layer(num_neurons_per_layer),
+    _num_hidden_layers(num_neurons_per_layer.size()),
+    _activation_function(MultiMooseEnum("relu sigmoid elu gelu linear", "relu"))
+{
+  _activation_function = activation_function;
+
+  // Check if the number of activation functions matches the number of hidden layers
+  if ((_activation_function.size() != 1) && (_activation_function.size() != _num_hidden_layers))
+    mooseError("The number of activation functions should be either one or the same as the number "
+               "of hidden layers");
+  constructNeuralNetwork();
+}
+
+void
+LibtorchArtificialNeuralNet::constructNeuralNetwork()
+{
+  // Adding hidden layers
+  unsigned int inp_neurons = _num_inputs;
+  for (unsigned int i = 0; i < _num_hidden_layers; ++i)
+  {
+    std::unordered_map<std::string, unsigned int> parameters = {
+        {"inp_neurons", inp_neurons}, {"out_neurons", _num_neurons_per_layer[i]}};
+    addLayer("hidden_layer_" + std::to_string(i + 1), parameters);
+
+    // Necessary to retain double precision (and error-free runs)
+    _weights[i]->to(at::kDouble);
+    inp_neurons = _num_neurons_per_layer[i];
+  }
+  // Adding output layer
+  std::unordered_map<std::string, unsigned int> parameters = {{"inp_neurons", inp_neurons},
+                                                              {"out_neurons", _num_outputs}};
+  addLayer("output_layer_", parameters);
+  _weights.back()->to(at::kDouble);
+}
+
+torch::Tensor
+LibtorchArtificialNeuralNet::forward(torch::Tensor x)
+{
+  for (unsigned int i = 0; i < _weights.size() - 1; ++i)
+  {
+    std::string activation =
+        _activation_function.size() > 1 ? _activation_function[i] : _activation_function[0];
+    if (activation == "relu")
+      x = torch::relu(_weights[i]->forward(x));
+    else if (activation == "sigmoid")
+      x = torch::sigmoid(_weights[i]->forward(x));
+    else if (activation == "elu")
+      x = torch::elu(_weights[i]->forward(x));
+    else if (activation == "gelu")
+      x = torch::gelu(_weights[i]->forward(x));
+    else if (activation == "linear")
+      x = _weights[i]->forward(x);
+  }
+
+  x = _weights[_weights.size() - 1]->forward(x);
+
+  return x.reshape({x.size(0)});
+}
+
+void
+LibtorchArtificialNeuralNet::addLayer(
+    const std::string layer_name, const std::unordered_map<std::string, unsigned int> & parameters)
```

Comment:
`const std::string & layer_name`

Reply:
It is the same, I would like to use lines like this:

`addLayer("hidden_layer_" + std::to_string(i + 1), parameters);`

So I'd like to keep this as is.

Reply:
I'll take this comment back too, looks like your suggestion works.